ui/StatusQ/src/writableproxymodel.cpp

```diff
@@ -0,0 +1,838 @@
+{
+    if(at < 0 || at >= rowCount())
+    {
+        return {};
+    }
+
+    auto index = this->index(at, 0);
+    auto data = this->itemData(index);
+    auto roleNames = this->roleNames();
+    QVariantMap rowMap;
+    for(auto it = data.begin(); it != data.end(); ++it)
+    {
+        auto roleName = roleNames[it.key()];
+        rowMap[roleName] = it.value();
+    }
+    return rowMap;
+}
+
+bool WritableProxyModel::set(int at, const QVariantMap& data)
+{
+    if(at < 0 || at >= rowCount())
+    {
+        return false;
+    }
+
+    auto index = this->index(at, 0);
+    auto itemData = this->itemData(index);
+    auto roleNames = this->roleNames();
+
+    for(auto it = data.begin(); it != data.end(); ++it)
+    {
+        auto role = roleNames.key(it.key().toUtf8(), -1);
+        itemData[role] = it.value();
+    }
+
+    return setItemData(index, itemData);
+}
+
+bool WritableProxyModel::dirty() const
+{
+    return m_dirty;
+}
+
+void WritableProxyModel::setDirty(bool flag)
+{
+    if(m_dirty == flag || !d->canUpdateDirtyFlag)
+    {
+        return;
+    }
+
+    m_dirty = flag;
+    emit dirtyChanged();
+}
+
+void WritableProxyModel::setSourceModel(QAbstractItemModel* sourceModel)
+{
+    if(sourceModel == QAbstractProxyModel::sourceModel())
+    {
+        return;
+    }
+
+    beginResetModel();
+
+    d->clear();
+
+    if(QAbstractProxyModel::sourceModel())
+    {
+        disconnect(QAbstractProxyModel::sourceModel(), nullptr, this, nullptr);
+    }
+
+    setDirty(false);
+    QAbstractProxyModel::setSourceModel(sourceModel);
+
+    if(!sourceModel)
+    {
+        endResetModel();
+        return;
+    }
+
+    d->applyRoleNamesFix();
+
+    d->createProxyToSourceRowMap();
+    connect(sourceModel, &QAbstractItemModel::dataChanged, this, &WritableProxyModel::handleSourceDataChanged);
+    connect(sourceModel, &QAbstractItemModel::rowsAboutToBeInserted, this, &WritableProxyModel::handleRowsAboutToBeInserted);
+    connect(sourceModel, &QAbstractItemModel::rowsInserted, this, &WritableProxyModel::handleRowsInserted);
+    connect(sourceModel, &QAbstractItemModel::rowsAboutToBeRemoved, this, &WritableProxyModel::handleRowsAboutToBeRemoved);
+    connect(sourceModel, &QAbstractItemModel::rowsRemoved, this, &WritableProxyModel::handleRowsRemoved);
+    connect(sourceModel, &QAbstractItemModel::modelAboutToBeReset, this, &WritableProxyModel::handleModelAboutToBeReset);
+    connect(sourceModel, &QAbstractItemModel::modelReset, this, &WritableProxyModel::handleModelReset);
+    connect(sourceModel, &QAbstractItemModel::rowsMoved, this, &WritableProxyModel::handleRowsMoved);
+    connect(sourceModel, &QAbstractItemModel::layoutAboutToBeChanged, this, &WritableProxyModel::handleLayoutAboutToBeChanged);
+    connect(sourceModel, &QAbstractItemModel::layoutChanged, this, &WritableProxyModel::handleModelReset);
+    
+    endResetModel();
+}
+
+int	WritableProxyModel::columnCount(const QModelIndex &parent) const
```

Comment:
```suggestion
int WritableProxyModel::columnCount(const QModelIndex &parent) const
```
some tabs instead of spaces in several places